File: internal/okf/frontmatter.go
// Package okf is the OKF substrate: a hand-written frontmatter splitter and
// glob matching. It knows nothing about the SDLC type taxonomy — it only deals
// with markdown + YAML frontmatter and bundle file discovery.
package okf

import (
	"errors"
	"fmt"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// lineEndingNormalizer collapses CRLF and lone-CR line endings to LF so the
// frontmatter splitter and YAML decoder see consistent input regardless of how
// a file was authored (classic Mac "\r", Windows "\r\n", or Unix "\n").
var lineEndingNormalizer = strings.NewReplacer("\r\n", "\n", "\r", "\n")

// ErrUnterminated is returned when a file opens a `---` fence but never closes it.
var ErrUnterminated = errors.New("frontmatter: opening --- without a closing fence")

// Document is a parsed OKF concept: its frontmatter mapping plus body.
type Document struct {
	Path           string
	HasFrontmatter bool
	Map            *yaml.Node // the frontmatter mapping node, nil when HasFrontmatter is false
	Body           string
	frontStart     int // 1-based file line of the first frontmatter content line
}

// LinkRef is one typed link target as written, with its source line.
type LinkRef struct {
	Raw  string
	Line int
}

// Split separates a leading `--- ... ---` YAML frontmatter block from the body.
// The opening fence must be the very first line (after an optional BOM). The
// closing fence is the next line that is exactly `---`. Returns ok=false (no
// error) when there is no opening fence, and ErrUnterminated when the opener has
// no closer.
func Split(raw []byte) (front []byte, body string, frontStartLine int, ok bool, err error) {
	s := strings.TrimPrefix(string(raw), "\ufeff") // strip UTF-8 BOM if present
	s = lineEndingNormalizer.Replace(s)            // normalize CRLF / lone CR to LF
	lines := strings.Split(s, "\n")
	if len(lines) == 0 || strings.TrimRight(lines[0], " \t\r") != "---" {
		return nil, s, 0, false, nil
	}
	for i := 1; i < len(lines); i++ {
		if strings.TrimRight(lines[i], " \t\r") == "---" {
			return []byte(strings.Join(lines[1:i], "\n")), strings.Join(lines[i+1:], "\n"), 2, true, nil
		}
	}
	return nil, s, 0, false, ErrUnterminated
}

// Parse splits and YAML-decodes a file. A document with no frontmatter is not an
// error (HasFrontmatter is false) — this is how reserved/prose files are tolerated.
func Parse(path string, raw []byte) (*Document, error) {
	front, body, start, ok, err := Split(raw)
	d := &Document{Path: path, Body: body, frontStart: start}
	if err != nil {
		return d, err
	}
	if !ok {
		return d, nil
	}
	d.HasFrontmatter = true
	var doc yaml.Node
	if e := yaml.Unmarshal(front, &doc); e != nil {
		return d, fmt.Errorf("frontmatter: %w", e)
	}
	switch {
	case len(doc.Content) == 0:
		d.Map = &yaml.Node{Kind: yaml.MappingNode} // empty frontmatter -> empty mapping
	case doc.Content[0].Kind == yaml.MappingNode:
		d.Map = doc.Content[0]
	default:
		return d, fmt.Errorf("frontmatter: expected a mapping, got %s", kindName(doc.Content[0].Kind))
	}
	return d, nil
}

// Line maps a YAML node (numbered relative to the frontmatter block) to a
// 1-based line in the original file. Returns 0 for a nil node.
func (d *Document) Line(n *yaml.Node) int {
	if n == nil || d.frontStart == 0 {
		return 0
	}
	return n.Line + d.frontStart - 1
}

// FrontStartLine is the file line of the first frontmatter content line (2 for a
// well-formed file), used as the fallback location for "missing key" findings.
func (d *Document) FrontStartLine() int { return d.frontStart }

// Get returns the key and value nodes for a top-level frontmatter key.
func (d *Document) Get(key string) (keyNode, valNode *yaml.Node, ok bool) {
	if d.Map == nil {
		return nil, nil, false
	}
	for i := 0; i+1 < len(d.Map.Content); i += 2 {
		if d.Map.Content[i].Value == key {
			return d.Map.Content[i], d.Map.Content[i+1], true
		}
	}
	return nil, nil, false
}

// Keys returns the top-level frontmatter keys in document order.
func (d *Document) Keys() []string {
	var out []string
	if d.Map != nil {
		for i := 0; i+1 < len(d.Map.Content); i += 2 {
			out = append(out, d.Map.Content[i].Value)
		}
	}
	return out
}

// Links reads the `links:` block: a mapping of relationship name to a list of
// (or single) root-relative path targets, each carrying its source line.
func (d *Document) Links() map[string][]LinkRef {
	res := map[string][]LinkRef{}
	_, lv, ok := d.Get("links")
	if !ok || lv.Kind != yaml.MappingNode {
		return res
	}
	for i := 0; i+1 < len(lv.Content); i += 2 {
		rel := lv.Content[i].Value
		val := lv.Content[i+1]
		var refs []LinkRef
		switch val.Kind {
		case yaml.SequenceNode:
			for _, item := range val.Content {
				if raw, line, ok := scalarRef(item, d); ok {
					refs = append(refs, LinkRef{Raw: raw, Line: line})
				}
				// Non-scalar items (mappings / nested sequences) are skipped:
				// appending a bogus empty Raw would be worse than omitting it,
				// and the now-shorter list trips the relationship's min check.
			}
		case yaml.ScalarNode:
			if val.Value != "" {
				refs = append(refs, LinkRef{Raw: val.Value, Line: d.Line(val)})
			}
		case yaml.AliasNode:
			if raw, line, ok := scalarRef(val, d); ok {
				refs = append(refs, LinkRef{Raw: raw, Line: line})
			}
		}
		res[rel] = refs
	}
	return res
}

// scalarRef resolves a sequence item to its scalar string value and source
// line. Alias nodes are dereferenced to the node they point at; only a scalar
// (the alias's target or the item itself) yields ok=true. Mappings, nested
// sequences, and empty scalars are rejected (ok=false) so the caller can skip
// them rather than emit a bogus link target.
func scalarRef(item *yaml.Node, d *Document) (raw string, line int, ok bool) {
	if item == nil {
		return "", 0, false
	}
	if item.Kind == yaml.AliasNode {
		if item.Alias == nil || item.Alias.Kind != yaml.ScalarNode || item.Alias.Value == "" {
			return "", 0, false
		}
		return item.Alias.Value, d.Line(item), true
	}
	if item.Kind == yaml.ScalarNode && item.Value != "" {
		return item.Value, d.Line(item), true
	}
	return "", 0, false
}

// DuplicateTopLevelKeys reports top-level frontmatter keys that appear more than
// once, sorted and deduped. YAML decoding keeps only the last value for a
// duplicated key in lookups, so callers use this to flag the silent shadowing.
// Tolerant: a nil Map yields nil.
func (d *Document) DuplicateTopLevelKeys() []string {
	if d.Map == nil {
		return nil
	}
	seen := map[string]int{}
	for i := 0; i+1 < len(d.Map.Content); i += 2 {
		seen[d.Map.Content[i].Value]++
	}
	return dupKeys(seen)
}

// DuplicateLinkRels reports relationship names that appear more than once inside
// the `links:` mapping, sorted and deduped. Tolerant: a nil Map or a missing /
// non-mapping `links:` yields nil.
func (d *Document) DuplicateLinkRels() []string {
	_, lv, ok := d.Get("links")
	if !ok || lv == nil || lv.Kind != yaml.MappingNode {
		return nil
	}
	seen := map[string]int{}
	for i := 0; i+1 < len(lv.Content); i += 2 {
		seen[lv.Content[i].Value]++
	}
	return dupKeys(seen)
}

// dupKeys returns the keys whose count exceeds one, sorted for determinism.
func dupKeys(counts map[string]int) []string {
	var out []string
	for k, n := range counts {
		if n > 1 {
			out = append(out, k)
		}
	}
	sort.Strings(out)
	return out
}

func kindName(k yaml.Kind) string {
	switch k {
	case yaml.SequenceNode:
		return "list"
	case yaml.ScalarNode:
		return "scalar"
	case yaml.MappingNode:
		return "mapping"
	default:
		return "value"
	}
}

File: src/core/okf/frontmatter.ts
// The OKF substrate: a hand-written frontmatter splitter over the `yaml` AST.
// Knows nothing about the SDLC taxonomy — only markdown + YAML frontmatter.
// Port of Go internal/okf/frontmatter.go.
//
// Line numbers: the `yaml` package gives byte ranges, not line numbers, so we
// parse the front block in isolation with a LineCounter and convert range[0] ->
// 1-based line, then re-base by frontStart (exactly as Go did node.Line+start-1).
// `uniqueKeys: false` keeps duplicate keys in document order so we can detect the
// silent-shadowing the validator warns about (Go walked the node Content slice).
import { parseDocument, LineCounter, isMap, isScalar, isSeq, isAlias, YAMLMap } from "yaml";
import type { Node, Document as YamlAst } from "yaml";

export class UnterminatedError extends Error {
  constructor() {
    super("frontmatter: opening --- without a closing fence");
    this.name = "UnterminatedError";
  }
}

export interface LinkRef {
  raw: string;
  line: number;
}

export interface SplitResult {
  front: string | null;
  body: string;
  frontStartLine: number;
  ok: boolean;
  err: Error | null;
}

function trimRightWS(s: string): string {
  return s.replace(/[ \t\r]+$/, "");
}

// split separates a leading `--- ... ---` YAML frontmatter block from the body.
// The opening fence must be the very first line (after an optional BOM). Returns
// ok=false (no error) when there is no opening fence, and an UnterminatedError
// when the opener has no closer.
export function split(raw: string | Uint8Array): SplitResult {
  let s = typeof raw === "string" ? raw : new TextDecoder().decode(raw);
  if (s.startsWith("﻿")) s = s.slice(1); // strip UTF-8 BOM
  s = s.replaceAll("\r\n", "\n").replaceAll("\r", "\n"); // normalize CRLF / lone CR
  const lines = s.split("\n");
  if (lines.length === 0 || trimRightWS(lines[0]!) !== "---") {
    return { front: null, body: s, frontStartLine: 0, ok: false, err: null };
  }
  for (let i = 1; i < lines.length; i++) {
    if (trimRightWS(lines[i]!) === "---") {
      return {
        front: lines.slice(1, i).join("\n"),
        body: lines.slice(i + 1).join("\n"),
        frontStartLine: 2,
        ok: true,
        err: null,
      };
    }
  }
  return { front: null, body: s, frontStartLine: 0, ok: false, err: new UnterminatedError() };
}

// Document is a parsed OKF concept: its frontmatter mapping plus body.
export class Document {
  readonly path: string;
  hasFrontmatter = false;
  body: string;
  private frontStart: number;
  private lc: LineCounter | null = null;
  private ast: YamlAst | null = null;
  map: YAMLMap | null = null; // frontmatter mapping node, null when hasFrontmatter is false

  constructor(path: string, body: string, frontStart: number) {
    this.path = path;
    this.body = body;
    this.frontStart = frontStart;
  }

  // line maps a YAML node to a 1-based line in the original file. 0 for nil.
  line(n: Node | null | undefined): number {
    if (!n || this.frontStart === 0 || !this.lc) return 0;
    const r = n.range;
    if (!r) return 0;
    return this.lc.linePos(r[0]).line + this.frontStart - 1;
  }

  // frontStartLine is the file line of the first frontmatter content line (2 for a
  // well-formed file) — the fallback location for "missing key" findings.
  frontStartLine(): number {
    return this.frontStart;
  }

  // get returns the key and value nodes for a top-level frontmatter key (first
  // occurrence, matching Go's Content walk).
  get(key: string): { keyNode: Node; valNode: Node } | undefined {
    if (!this.map) return undefined;
    for (const pair of this.map.items) {
      const k = pair.key as Node;
      if (isScalar(k) && String(k.value) === key) {
        return { keyNode: k, valNode: pair.value as Node };
      }
    }
    return undefined;
  }

  // keys returns the top-level frontmatter keys in document order.
  keys(): string[] {
    if (!this.map) return [];
    return this.map.items.map((p) => String((p.key as { value?: unknown }).value));
  }

  // toJS converts the frontmatter mapping to a plain JS object (decoded values).
  toJS(): Record<string, unknown> {
    if (!this.map || this.map.items.length === 0) return {};
    return this.map.toJS(this.ast!) as Record<string, unknown>;
  }

  private scalarRef(item: Node | null): LinkRef | null {
    if (!item) return null;
    if (isAlias(item)) {
      const target = item.resolve(this.ast!);
      if (!target || !isScalar(target) || target.value === "" || target.value == null) return null;
      return { raw: String(target.value), line: this.line(item) };
    }
    if (isScalar(item) && item.value !== "" && item.value != null) {
      return { raw: String(item.value), line: this.line(item) };
    }
    return null;
  }

  // links reads the `links:` block: relationship name -> list of root-relative
  // path targets, each carrying its source line.
  links(): Record<string, LinkRef[]> {
    const res: Record<string, LinkRef[]> = {};
    const g = this.get("links");
    if (!g || !isMap(g.valNode)) return res;
    for (const pair of g.valNode.items) {
      const rel = String((pair.key as { value?: unknown }).value);
      const val = pair.value as Node;
      const refs: LinkRef[] = [];
      if (isSeq(val)) {
        for (const item of val.items) {
          const sr = this.scalarRef(item as Node);
          if (sr) refs.push(sr);
        }
      } else if (isScalar(val)) {
        if (val.value !== "" && val.value != null) refs.push({ raw: String(val.value), line: this.line(val) });
      } else if (isAlias(val)) {
        const sr = this.scalarRef(val);
        if (sr) refs.push(sr);
      }
      res[rel] = refs;
    }
    return res;
  }

  duplicateTopLevelKeys(): string[] {
    if (!this.map) return [];
    return dupKeys(this.map.items.map((p) => String((p.key as { value?: unknown }).value)));
  }

  duplicateLinkRels(): string[] {
    const g = this.get("links");
    if (!g || !isMap(g.valNode)) return [];
    return dupKeys(g.valNode.items.map((p) => String((p.key as { value?: unknown }).value)));
  }

  // _attach wires the parsed AST + line counter (used by parse()).
  _attach(ast: YamlAst, lc: LineCounter, map: YAMLMap): void {
    this.ast = ast;
    this.lc = lc;
    this.map = map;
    this.hasFrontmatter = true;
  }
}

// --- node accessors (keep the `yaml` dependency localized to okf) ---

export type ValueNode = Node;

export function nodeIsScalar(n: Node | null | undefined): boolean {
  return !!n && isScalar(n);
}

export function nodeIsSeq(n: Node | null | undefined): boolean {
  return !!n && isSeq(n);
}

export function seqItems(n: Node | null | undefined): Node[] {
  return n && isSeq(n) ? (n.items as Node[]) : [];
}

// scalarText returns a scalar's value with Go's yaml.Node.Value parity: the
// unquoted string content for strings, and the raw source token for
// numbers/bools/dates (so a number field sees the text the author wrote). Empty
// string for a non-scalar, empty, or null node.
export function scalarText(n: Node | null | undefined): string {
  if (!n || !isScalar(n)) return "";
  const v = n.value;
  if (typeof v === "string") return v;
  if (v == null) return "";
  const src = (n as { source?: unknown }).source;
  return typeof src === "string" ? src : String(v);
}

// isEmptyScalar reports a present-but-empty scalar (blank value or explicit
// null) — does not satisfy a required field.
export function isEmptyScalar(n: Node | null | undefined): boolean {
  return !!n && isScalar(n) && (n.value === "" || n.value == null);
}

function dupKeys(names: string[]): string[] {
  const counts = new Map<string, number>();
  for (const n of names) counts.set(n, (counts.get(n) ?? 0) + 1);
  const out: string[] = [];
  for (const [k, n] of counts) if (n > 1) out.push(k);
  out.sort();
  return out;
}

function kindName(n: Node): string {
  if (isSeq(n)) return "list";
  if (isScalar(n)) return "scalar";
  if (isMap(n)) return "mapping";
  return "value";
}

// parse splits and YAML-decodes a file. A document with no frontmatter is not an
// error (hasFrontmatter is false) — this is how reserved/prose files are tolerated.
// Returns [doc, err]; err mirrors Go's (d, error) return for a malformed block.
export function parse(path: string, raw: string | Uint8Array): [Document, Error | null] {
  const sr = split(raw);
  const d = new Document(path, sr.body, sr.frontStartLine);
  if (sr.err) return [d, sr.err];
  if (!sr.ok || sr.front === null) return [d, null];

  const lc = new LineCounter();
  const ast = parseDocument(sr.front, { lineCounter: lc, uniqueKeys: false });
  const fatal = ast.errors[0];
  if (fatal) return [d, new Error(`frontmatter: ${fatal.message}`)];

  const contents = ast.contents;
  if (contents == null) {
    d._attach(ast, lc, new YAMLMap()); // empty frontmatter -> empty mapping
  } else if (isMap(contents)) {
    d._attach(ast, lc, contents);
  } else {
    return [d, new Error(`frontmatter: expected a mapping, got ${kindName(contents as Node)}`)];
  }
  return [d, null];
}
